src/string_with_vars.ml

```diff
@@ -1,111 +1,129 @@
+   files *)
+module Jbuild : sig
+  val parse : string -> loc:Loc.t -> quoted:bool -> t
+end = struct
+  type var_syntax = Parens | Braces
+  module Token = struct
+    type t =
+      | String of string
+      | Open   of var_syntax
+      | Close  of var_syntax
+
+    let tokenise s =
+      let len = String.length s in
+      let sub i j = String.sub s ~pos:i ~len:(j - i) in
+      let cons_str i j acc = if i = j then acc else String (sub i j) :: acc in
+      let rec loop i j =
+        if j = len
+        then cons_str i j []
+        else
+          match s.[j] with
+          | '}' -> cons_str i j (Close Braces :: loop (j + 1) (j + 1))
+          | ')' -> cons_str i j (Close Parens :: loop (j + 1) (j + 1))
+          | '$' when j + 1 < len -> begin
+              match s.[j + 1] with
+              | '{' -> cons_str i j (Open Braces :: loop (j + 2) (j + 2))
+              | '(' -> cons_str i j (Open Parens :: loop (j + 2) (j + 2))
+              | _   -> loop i (j + 1)
+            end
+          | _ -> loop i (j + 1)
+      in
+      loop 0 0
+
+    let to_string = function
+      | String s     -> s
+      | Open  Braces -> "${"
+      | Open  Parens -> "$("
+      | Close Braces -> "}"
+      | Close Parens -> ")"
+  end
+  (* Remark: Consecutive [Text] items are concatenated. *)
+  let rec of_tokens
+    : Loc.t -> Token.t list -> part list = fun loc -> function
+    | [] -> []
+    | Open a :: String s :: Close b :: rest when a = b ->
+      let (name, payload) =
+        match String.lsplit2 s ~on:':' with
+        | None -> ("", s)
+        | Some v -> v
+      in
+      Var { loc
+          ; name
+          ; payload
+          ; syntax =
+              begin match a with
+              | Parens -> Dollar_paren
+              | Braces -> Dollar_brace
+              end
+          } :: of_tokens loc rest
+    | token :: rest ->
+      let s = Token.to_string token in
+      match of_tokens loc rest with
+      | Text s' :: l -> Text (s ^ s') :: l
+      | l -> Text s :: l
+
+  let parse s ~loc ~quoted =
+    { parts = of_tokens loc (Token.tokenise s)
+    ; loc
+    ; quoted
+    }
 end
 
-(* Remark: Consecutive [Text] items are concatenated. *)
-let rec of_tokens : Token.t list -> item list = function
-  | [] -> []
-  | Open a :: String s :: Close b :: rest when a = b ->
-    Var (a, s) :: of_tokens rest
-  | token :: rest ->
-    let s = Token.to_string token in
-    match of_tokens rest with
-    | Text s' :: l -> Text (s ^ s') :: l
-    | l -> Text s :: l
-
-let items_of_string s = of_tokens (Token.tokenise s)
-
 let t =
   let open Sexp.Of_sexp in
-  raw >>| fun sexp ->
-  match sexp with
-  | Atom(loc, A s) -> { items = items_of_string s;  loc;  quoted = false }
-  | Quoted_string (loc, s) ->
-    { items = items_of_string s;  loc;  quoted = true }
-  | List (loc, _) -> of_sexp_error loc "Atom or quoted string expected"
+  Syntax.get_exn Stanza.syntax >>= function
+  | (0, _) ->
+    begin raw >>| function
```

Comment:
What do you think about extracting some function for these two branches? In particular, this one could go in a `Jbuild.t` value.

Reply:
Yeah, that's cleaner.